### ISSUE TITLE
Assertion failure triggerable after attempting to add duplicate foreign pages

### ISSUE THREAD

**Issue by jbarlow83** [qpdf/qpdf#514]
I came across this assertion failure while trying to get QPDF to accept adding the same foreign page twice to a file. (Suppose the page is a separator we're adding between sections of a document.)

I managed to trigger the following assertion failure:

```
Assertion failed: (this->m->pageobj_to_pages_pos.size() == QIntC::to_size(npages)), function insertPage, file libqpdf/QPDF_pages.cc, line 264.
```

In short it seems to me that after a failure to add the same foreign page twice, some internal state is left inconsistent.

The following pikepdf code will reproduce the issue:

```python
import pikepdf

g = pikepdf.open('tests/resources/graph.pdf')
p = pikepdf.new()

p.pages.append(g.pages[0])

try:
    p.pages.append(g.pages[0])  # Raises an exception - duplicate page reference
except pikepdf.PdfError:
    print("Hit the expected exception")

foreign = p.copy_foreign(g.pages[0])
print("We're going to SIGABRT now...")
p.pages.append(foreign)

```

Or roughly equivalent C++:

```c++
#include <iostream>

#include <qpdf/QPDF.hh>
#include <qpdf/QPDFExc.hh>
#include <qpdf/QPDFSystemError.hh>
#include <qpdf/QPDFObjGen.hh>
#include <qpdf/QPDFXRefEntry.hh>
#include <qpdf/Buffer.hh>
#include <qpdf/BufferInputSource.hh>
#include <qpdf/QPDFWriter.hh>
#include <qpdf/QPDFPageDocumentHelper.hh>
#include <qpdf/Pl_Discard.hh>


int main (int argc, char **argv) {
    QPDF src;
    src.processFile(argv[1]);

    QPDF dst;
    dst.emptyPDF();

    auto page0 = src.getAllPages().at(0);

    dst.addPage(page0, false);

    try {
        dst.addPage(page0, false); 
    } catch (const QPDFExc& e) {
        std::cout << "Expected exception:" << std::endl;
        std::cout << e.what() << std::endl;
    }

    std::cout << "Make foreign:" << std::endl;
    auto foreign = dst.copyForeignObject(page0);
    auto foreign_copy = foreign.shallowCopy();
    std::cout << "Assertion failure happens here:" << std::endl;
    dst.addPage(foreign_copy, false);
    return 0;
}

```

(And I unlike the last time I reported such a similar issue, I made sure the C++ compiles and does what I think it does rather than try to simplify in the GitHub editor. Sorry about that....)

I compiled the code above on macOS with the following, linked against qpdf 10.3.1:
```bash
clang++ -std=c++14 -o assert_issue assert_issue.cpp -lqpdf
```

To execute:
```bash
./assert_issue ../tests/resources/graph.pdf
```

The output should look like:
```
Expected exception:
empty PDF (page 1 (numbered from zero): object 3 0): duplicate page reference found; this would cause loss of data
Make foreign:
Assertion failure happens here:
Assertion failed: (this->m->pageobj_to_pages_pos.size() == QIntC::to_size(npages)), function insertPage, file libqpdf/QPDF_pages.cc, line 264.
fish: 'env TC_SCOPE=qpdf TC_FILENAME=f…' terminated by signal SIGABRT (Abort)
```

The second `.addPage()`, which raises an exception, is essential to triggering this issue, as is the `.shallowCopy()`. Somehow doing this seems to create some inconsistency, presumably in `pageobj_to_pages_pos`.

**Comment by jberkenbilt** [comment]
I'll study this and see what I can figure out. This part of the qpdf library is one of the more complex bits, and there are things about the implementation and the interface that I would definitely do differently if I were to do it over. The pageobj_to_pages stuff is one of the very few bits of the library that was written by someone else (a very talented student who did great work overall), and all the page caching and flattening page tree stuff has been a liability I've been tempted to remove. The big problem is figuring out what I'm going to do about getAllPages(), which is the worst bit of interface in all of qpdf as returns a writable reference to an internal cache. I have considered breaking that, but it would be the first interface change in the history of qpdf's existence as an open source project that would possibly require people to change some logic around if anyone has written code that relies on changes to the pages structures reflecting in that value....

Anyway, I'll see what I can do. Thanks for the report.

**Comment by jberkenbilt** [comment]
Soon I plan on creating a higher level interface for working with pages and document-level data that ties to pages like form fields and outlines. The current interfaces are very hard to get right, and some early decisions I made about implementation make it hard to fix without incurring significant runtime overhead. Anyway, I'm just speculating. I'll have to actually dig into this.

**Comment by jberkenbilt** [comment]
Looking back at the code, I have to take 100% of the blame, but according to my own comments, the method that returns the const reference to the internal cache actually predates addition of any methods that could have mutated the state. qpdf started its life as a read-only PDF library.

The pages cache is not used very heavily, and I think I can rework the code to just eliminate it. In the meantime, the issue is that, because there are two pages with the same object, the arrays in question actually have different sizes. This is known in the code because of the earlier warning.

Bottom line: the fix here appears to be simply removing the assertions. I long ago stopped creating assertions in the library code and had eliminated most of them. assertions belong in test code. In this case, the assertion failure is simply a side effect of the already-reported data loss, but it doesn't necessarily result in data loss. If you add a write, you get a PDF with three pages, two of which share the same ID. Not ideal, but it renders properly at least with some viewers.

**Comment by jberkenbilt** [comment]
I started reworking the page handling code internally and, as I did, I came to a deeper understanding of why it's implemented the way it is and actually think it's probably pretty much correct. I'll update comments and notes so I don't again in the future decide that the existing code is broken. In any case, whatever the next qpdf release is will automatically resolve duplicates when inserting new pages and will no longer throw an exception if you call addPage more than once.

**Comment by jberkenbilt** [comment]
I'm still planning on the higher level interface I mentioned above, just not on getting rid of the caches.

**Comment by jberkenbilt** [comment]
For reference, the other commit is 8971443e4680fc1c0babe56da58cc9070a9dae2e
